Sampling Output checks and Normality tests to check for bugs

springcoil opened the pull request:
Based on https://github.com/pymc-devs/pymc3/pull/1424 - I couldn't for some reason rebase that correctly - added in some KS-tests etc. 

Needs a review.

twiecki commented on pymc3/tests/test_step.py:
```diff
@@ -39,26 +41,26 @@ class TestStepMethods(object):  # yield test doesn't work subclassing unittest.T
             7.04959179e-01, 8.37863464e-01, -5.24200836e-01, 1.28261340e+00, 9.08774240e-01,
             8.80566763e-01, 7.82911967e-01, 8.01843432e-01, 7.09251098e-01, 5.73803618e-01]),
         HamiltonianMC: np.array([
-           -0.74925631, -0.2566773 , -2.12480977,  1.64328926, -1.39315913,
```
Can you remove these from the diff?

twiecki commented on pymc3/tests/test_step.py:
```diff
@@ -238,3 +240,37 @@ def test_binomial(self):
+                assert np.isclose(np.median(trace.beta, 0), beta_true, rtol=0.1).all()
+                assert np.isclose(np.median(trace.alpha), alpha_true, rtol=0.1)
+                assert np.isclose(np.median(trace.sigma), sigma_true, rtol=0.1)
+                np.random.seed(987654321)
```
Since we inherit from SeededTest, we shouldn't need to set it again here. Sorry for being confusing.

twiecki commented on pymc3/tests/test_step.py:
```diff
@@ -238,3 +240,37 @@ def test_binomial(self):
+                test_normal = stats.kstest(trace.alpha, 'norm', alternative='greater')
+                test_normal_beta = stats.kstest(trace.beta[0], 'norm', alternative='greater')
+                test_uniform = stats.kstest(trace.sigma, 'uniform', alternative='greater')
+                assert np.less(np.median(test_normal[1]), 0.05)
```
What are the p-values? Could we make the test more stringent, perhaps by increasing the number of samples?

Also, I think https://docs.scipy.org/doc/numpy-1.10.0/reference/generated/numpy.testing.assert_array_less.html is better here.

twiecki commented on pymc3/tests/test_step.py:
```diff
@@ -238,3 +240,37 @@ def test_binomial(self):
+                assert np.isclose(np.median(trace.alpha), alpha_true, rtol=0.1)
+                assert np.isclose(np.median(trace.sigma), sigma_true, rtol=0.1)
+                np.random.seed(987654321)
+                test_normal = stats.kstest(trace.alpha, 'norm', alternative='greater')
```
how about: `_, p_normal = ...`, can then drop the index below.

twiecki commented on pymc3/tests/test_step.py:
```diff
@@ -238,3 +240,37 @@ def test_binomial(self):
+                test_normal_beta = stats.kstest(trace.beta[0], 'norm', alternative='greater')
+                test_uniform = stats.kstest(trace.sigma, 'uniform', alternative='greater')
+                assert np.less(np.median(test_normal[1]), 0.05)
+                assert np.less(np.median(test_normal_beta[1]) / 2, 0.5)
```
why `/ 2`? and .5?

springcoil replied:
That was to hack it together. I'm fixing this now. Thanks for the feedback.

springcoil committed:
Fixing up tests

Updating tests

Updating tests again

Added in one test for each sampler

Updating tests

Sampling errors

Trying this commit

Removing parameterized nose tests as they didn't work with the model
context

Moving asserts and removing arguments which are unnecessary - removing requirements.txt reference to nose_parameterized

Changing shape

Removing unnecessary test, and checking they run locally

springcoil commented:
@twiecki Thanks for the feedback. I changed the tests slightly based on your suggestions, thanks. The tests are now hopefully a lot more robust. We may have to relax the conditions slightly, but they passed on my computer.

twiecki commented on pymc3/tests/test_step.py:
```diff
@@ -238,3 +239,36 @@ def test_binomial(self):
+                                [Slice([alpha, sigma]), Metropolis([beta])]):
+                trace = sample(1000, step=step_method, progressbar=False)
+
+                assert np.isclose(np.median(trace.beta, 0), beta_true, rtol=0.1).all()
```
np.testing.assert_array_almost_equal()

twiecki commented on pymc3/tests/test_step.py:
```diff
@@ -238,3 +239,36 @@ def test_binomial(self):
+                _, test_normal = stats.kstest(trace.alpha, 'norm', alternative='greater')
+                _, test_normal_beta = stats.kstest(trace.beta[0], 'norm', alternative='greater')
+                _, test_uniform = stats.kstest(trace.sigma, 'uniform', alternative='greater')
+                np.testing.assert_array_almost_equal(np.median(test_normal), 4.9960036108132044e-15, decimal=2)
```
what is this testing? shouldn't those be just p-values?

springcoil replied:
Hmm maybe something went wrong. Let me find out...

springcoil replied:
Actually i'm using the wrong variant of the test, https://docs.scipy.org/doc/scipy-0.14.0/reference/generated/scipy.stats.kstest.html - must have gotten mixed up in my experiments.

twiecki commented on pymc3/tests/test_step.py:
```diff
@@ -238,3 +239,36 @@ def test_binomial(self):
+                assert np.isclose(np.median(trace.sigma), sigma_true, rtol=0.1)
+                _, test_normal = stats.kstest(trace.alpha, 'norm', alternative='greater')
+                _, test_normal_beta = stats.kstest(trace.beta[0], 'norm', alternative='greater')
+                _, test_uniform = stats.kstest(trace.sigma, 'uniform', alternative='greater')
```
Shouldn't this also test for a normal distribution?

springcoil replied:
sigma = Uniform('sigma', lower=0.0, upper=1.0) - surely I need to test for the uniform distribtion? Or am i mistaken..n

twiecki replied:
But you're testing the posterior which should be Normal-ish. Log-normal I guess. Maybe we'll leave that one out here.

springcoil replied:
Ok fair point :) I'll add in something normalish.

springcoil commented:
There were some errors caused during my rebase. Fixing now...

springcoil commented:
Having trouble getting the p-values to work as we intended... seems to be quite tricky.

springcoil commented:
I'm having trouble diagnosing this error. I'll try to write up something though - even though it's over the holidays and I'm sleep deprived.

springcoil commented:
Ahh I think I get it. From the `scipy.stats` documentation
```python
*Testing t distributed random variables against normal distribution*

With 100 degrees of freedom the t distribution looks close to the normal
distribution, and the K-S test does not reject the hypothesis that the
sample came from the normal distribution:

>>> np.random.seed(987654321)
>>> stats.kstest(stats.t.rvs(100,size=100),'norm')
(0.072018929165471257, 0.67630062862479168)

With 3 degrees of freedom the t distribution looks sufficiently different
from the normal distribution, that we can reject the hypothesis that the
sample came from the normal distribution at the 10% level:

>>> np.random.seed(987654321)
>>> stats.kstest(stats.t.rvs(3,size=100),'norm')
(0.131016895759829, 0.058826222555312224)
```
So this is testing if the two distributions are different, the p-value being below say 10% proves that they are different distributions - not if they are the same distribution. Does anyone know another test or am I misunderstanding this?

twiecki commented:
Ohh, right, I forgot about that. How high are the p-values?

springcoil commented:
I'll include the diagnostics later - boarding a plane atm :)

springcoil commented:
`NormaltestResult(statistic=288.8868099649228, pvalue=1.8579168299255905e-63)` is the result from the Normal test for `stats.normaltest(trace.beta[:,1])` 

And ```(Pdb) stats.normaltest(trace.sigma)
NormaltestResult(statistic=31.217980013036858, pvalue=1.6638024983177679e-07)```

Submitting an updated PR now.

springcoil commented:
@twiecki Ready to merge?

twiecki commented on pymc3/tests/test_step.py:
```diff
@@ -238,3 +240,36 @@ def test_binomial(self):
+                # See `https://github.com/scipy/scipy/blob/v0.18.1/scipy/stats/stats.py#L1352`
+                _, test_normal_beta = stats.normaltest(trace.beta[:,1])
+                _, test_uniform = stats.normaltest(trace.sigma)
+                np.testing.assert_array_less(test_normal_beta, 0.005)
```
But I thought we realized that these should *not* be significant, so this means the posteriors are *not* normal, indicating a problem with our sampling. CC @fonnesbeck @jsalvatier

twiecki replied:
From https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.normaltest.html: "This function tests the null hypothesis that a sample comes from a normal distribution."

springcoil commented:
I think this test indicates it is normal. because this is a different test.
nevertheless can someone else check this? @colcarrol @fonnesbeck.

On 26 Dec 2016 6:13 AM, "Thomas Wiecki" <notifications@github.com> wrote:

> *@twiecki* commented on this pull request.
> ------------------------------
>
> In pymc3/tests/test_step.py
> <https://github.com/pymc-devs/pymc3/pull/1618#pullrequestreview-14378349>:
>
> > +            sigma = Uniform('sigma', lower=0.0, upper=1.0)
> +            mu = alpha + beta[0] * X1 + beta[1] * X2
> +            Y_obs = Normal('Y_obs', mu=mu, sd=sigma, observed=Y)
> +
> +            for step_method in (NUTS(), Metropolis(),
> +                                [Slice([alpha, sigma]), Metropolis([beta])]):
> +                trace = sample(1000, step=step_method, progressbar=False)
> +
> +                np.testing.assert_array_almost_equal(np.median(trace.beta, 0), beta_true, decimal=1)
> +                np.testing.assert_array_almost_equal(np.median(trace.alpha), alpha_true, decimal=1)
> +                np.testing.assert_array_almost_equal(np.median(trace.sigma), sigma_true, decimal=1)
> +                # Using the Normal test from SciPy `scipy.stats.normaltest`
> +                # See `https://github.com/scipy/scipy/blob/v0.18.1/scipy/stats/stats.py#L1352` <https://github.com/scipy/scipy/blob/v0.18.1/scipy/stats/stats.py#L1352>
> +                _, test_normal_beta = stats.normaltest(trace.beta[:,1])
> +                _, test_uniform = stats.normaltest(trace.sigma)
> +                np.testing.assert_array_less(test_normal_beta, 0.005)
>
> But I thought we realized that these should *not* be significant, so this
> means the posteriors are *not* normal, indicating a problem with our
> sampling. CC @fonnesbeck <https://github.com/fonnesbeck> @jsalvatier
> <https://github.com/jsalvatier>
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/pymc-devs/pymc3/pull/1618#pullrequestreview-14378349>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA8DiGpxmBaofOsuSw9Lctj-UUtl2jRzks5rL1r-gaJpZM4LSb5I>
> .
>

springcoil commented:
Ok we've got problems then with the samplers. These tests are worth doing.

Good point.

On 26 Dec 2016 12:56 PM, "Thomas Wiecki" <notifications@github.com> wrote:

> *@twiecki* commented on this pull request.
> ------------------------------
>
> In pymc3/tests/test_step.py <https://github.com/pymc-devs/pymc3/pull/1618>
> :
>
> > +            sigma = Uniform('sigma', lower=0.0, upper=1.0)
> +            mu = alpha + beta[0] * X1 + beta[1] * X2
> +            Y_obs = Normal('Y_obs', mu=mu, sd=sigma, observed=Y)
> +
> +            for step_method in (NUTS(), Metropolis(),
> +                                [Slice([alpha, sigma]), Metropolis([beta])]):
> +                trace = sample(1000, step=step_method, progressbar=False)
> +
> +                np.testing.assert_array_almost_equal(np.median(trace.beta, 0), beta_true, decimal=1)
> +                np.testing.assert_array_almost_equal(np.median(trace.alpha), alpha_true, decimal=1)
> +                np.testing.assert_array_almost_equal(np.median(trace.sigma), sigma_true, decimal=1)
> +                # Using the Normal test from SciPy `scipy.stats.normaltest`
> +                # See `https://github.com/scipy/scipy/blob/v0.18.1/scipy/stats/stats.py#L1352` <https://github.com/scipy/scipy/blob/v0.18.1/scipy/stats/stats.py#L1352>
> +                _, test_normal_beta = stats.normaltest(trace.beta[:,1])
> +                _, test_uniform = stats.normaltest(trace.sigma)
> +                np.testing.assert_array_less(test_normal_beta, 0.005)
>
> From https://docs.scipy.org/doc/scipy/reference/generated/
> scipy.stats.normaltest.html: "This function tests the null hypothesis
> that a sample comes from a normal distribution."
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/pymc-devs/pymc3/pull/1618>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA8DiJqdCiWulwDuZ7bT4I2G2sK0Tt-bks5rL7l7gaJpZM4LSb5I>
> .
>

twiecki commented on pymc3/tests/test_step.py:
```diff
@@ -238,3 +240,36 @@ def test_binomial(self):
+
+            for step_method in (NUTS(), Metropolis(),
+                                [Slice([alpha, sigma]), Metropolis([beta])]):
+                trace = sample(1000, step=step_method, progressbar=False)
```
I would draw more samples and add generous burn-in.

springcoil replied:
I'll add more generous burn-in.

twiecki commented on pymc3/tests/test_step.py:
```diff
@@ -238,3 +240,36 @@ def test_binomial(self):
+                # Using the Normal test from SciPy `scipy.stats.normaltest`
+                # See `https://github.com/scipy/scipy/blob/v0.18.1/scipy/stats/stats.py#L1352`
+                _, test_normal_beta = stats.normaltest(trace.beta[:,1])
+                _, test_uniform = stats.normaltest(trace.sigma)
```
Again, I would test for alpha, not sigma.

springcoil replied:
Thanks I'll do that.

springcoil commented:
Refactored this into two tests - one for NUTS and one for metropolis, added in Burn-in, a longer sampler and changed the statistical test. I'll push in a minute or two. 

These tests pass when the sampler is longer etc. Let me know if you need anything else in this. 

Update - Tests pass.

twiecki commented on pymc3/tests/test_step.py:
```diff
@@ -238,3 +241,129 @@ def test_binomial(self):
+            mu = alpha + beta[0] * X1 + beta[1] * X2
+            Y_obs = Normal('Y_obs', mu=mu, sd=sigma, observed=Y)
+
+            for step_method in (NUTS(), Metropolis(),
```
Not sure why you broke up the tests below?

springcoil replied:
I can put them back together if you want. I felt it might be good to have them broken up, so that we can understand what breaking changes do what. Metropolis is different to NUTS.

twiecki replied:
The proper way then would be to use `paremetrized.expand` which is quite elegant. See g.g. https://github.com/quantopian/alphalens/blob/master/alphalens/tests/test_performance.py#L60 In any case, should just be one test.

springcoil replied:
Ohh that looks neat :)

twiecki commented on pymc3/tests/test_step.py:
```diff
@@ -238,3 +241,129 @@ def test_binomial(self):
+            # We define very small as 0.05, for the sake of this argument.
+            # We test this by evaluating if the p-value is above 0.05.
+            _, test_normal = stats.normaltest(trace[-3000:].alpha)
+            np.testing.assert_array_less(0.05, test_normal, verbose=True)
```
So this tests that `.05 < p_normal` which is what we want, so we get a nice, normal posterior. I would add this test for alpha, beta_0 and beta_1.

springcoil replied:
I'll add that.

twiecki replied:
I'd also rename `test_normal` to `p_normal`.

springcoil replied:
Good suggestion :)

springcoil commented:
I made some commits - adding nose_parameterized stopped me being able to test it locally in my current conda environment. However it should work on travis.

fonnesbeck commented:
This is the last PR with a 3.0 milestone. Once this is done, I will do one more RC and (hopefully) release if all goes well.

springcoil commented:
Do we need another rc or can we go straight to a release?

twiecki commented:
We did do some fixes so probably best to do another RC.

twiecki commented on pymc3/tests/test_step.py:
```diff
@@ -238,3 +242,73 @@ def test_binomial(self):
+                                [Slice([alpha, sigma]), Metropolis([beta])]):
+                trace = sample(100000, step=step_method, progressbar=False)
+
+                np.testing.assert_array_almost_equal(np.median(trace.beta, 0), beta_true, decimal=1)
```
Why not move these tests into the one below and remove this whole test?

twiecki replied:
Once we remove this test I think this is ready.

twiecki commented on pymc3/tests/test_step.py:
```diff
@@ -238,3 +242,73 @@ def test_binomial(self):
+            # We define very small as 0.05, for the sake of this argument.
+            # We test this by evaluating if the p-value is above 0.05.
+            _, test_p = stats.normaltest(trace[-3000:].alpha)
+            _, test_p_beta_0 = stats.normaltest(trace[-3000:].beta[0])
```
Does that return an array or just a single sample? Thought it would have to be something like `.beta[0, :]` or some such.

springcoil replied:
Good point.

twiecki commented on pymc3/tests/test_step.py:
```diff
@@ -3,16 +3,20 @@
+from pymc3.distributions import Binomial, Normal, Bernoulli, Categorical, Uniform
 
 from numpy.testing import assert_array_almost_equal
+from nose_parameterized import parameterized
```
Need to add `nose_parameterized` to travis and dependencies. Or just revert to the list...

springcoil replied:
I added it to Travis and the dependencies. Getting context errors though.

springcoil replied:
Added it to the dependencies, for future use even though I'm probably not going to use it for this. Perhaps we should remove it though

springcoil replied:
I chose to use a list instead.

twiecki commented on requirements.txt:
```diff
@@ -11,3 +11,4 @@ recommonmark
 sphinx
 nbsphinx
 numpydoc
+nose_parameterized
```
don't need this anymore

springcoil commented:
Is this older test not needed then?

On 30 Dec 2016 10:29 PM, "Thomas Wiecki" <notifications@github.com> wrote:

> *@twiecki* commented on this pull request.
> ------------------------------
>
> In pymc3/tests/test_step.py <https://github.com/pymc-devs/pymc3/pull/1618>
> :
>
> > +        X1 = np.random.randn(size)
> +        X2 = np.random.randn(size) * 0.2
> +        Y = alpha_true + beta_true[0] * X1 + beta_true[1] * X2 + np.random.randn(size) * sigma_true
> +
> +        with Model() as model:
> +            alpha = Normal('alpha', mu=0, sd=10)
> +            beta = Normal('beta', mu=0, sd=10, shape=2)
> +            sigma = Uniform('sigma', lower=0.0, upper=1.0)
> +            mu = alpha + beta[0] * X1 + beta[1] * X2
> +            Y_obs = Normal('Y_obs', mu=mu, sd=sigma, observed=Y)
> +
> +            for step_method in (NUTS(), Metropolis(),
> +                                [Slice([alpha, sigma]), Metropolis([beta])]):
> +                trace = sample(100000, step=step_method, progressbar=False)
> +
> +                np.testing.assert_array_almost_equal(np.median(trace.beta, 0), beta_true, decimal=1)
>
> Once we remove this test I think this is ready.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/pymc-devs/pymc3/pull/1618>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA8DiOa2QR3ucCf7uvtjwie7GSu0kfNoks5rNYXWgaJpZM4LSb5I>
> .
>

twiecki commented:
It is but you can just move the asserts inside the other one.

springcoil commented:
Ok I added those changes, moved the asserts - removed an argument that wasn't needed anymore and removed references to nose_parameterized. We should (if all the tests pass) be ready to go.

twiecki commented on pymc3/tests/test_step.py:
```diff
@@ -238,3 +241,49 @@ def test_binomial(self):
+                # We test this by evaluating if the p-value is above 0.05.
+                _, test_p = stats.normaltest(trace[-3000:].alpha)
+                _, test_p_beta_0 = stats.normaltest(trace[-3000:].beta[0, :])
+                _, test_p_beta_1 = stats.normaltest(trace[-3000:].beta[:, 1])
```
They should both be sliced along the same dimension. What's the shape of beta?

springcoil replied:
Yeah oops - I meant to change that.

springcoil commented:
Here you go. This should be mergeable now.

springcoil commented:
Ok the tests pass I'm going to close the other PR that's related.

twiecki committed:
TST Improve test.

twiecki commented:
It was quicker for me to do it. I think this is what it should look like, but it doesn't pass yet for Slice.

springcoil commented:
Yeah it's fair that Slice should be there too. Maybe Slice applied to alpha
and beta only?

On 31 Dec 2016 5:00 PM, "Thomas Wiecki" <notifications@github.com> wrote:

> It was quicker for me to do it. I think this is what it should look like,
> but it doesn't pass yet for Slice.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/pymc-devs/pymc3/pull/1618#issuecomment-269873461>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AA8DiO3-IMNNG0DAcLYYxGZ4L0nBm_uyks5rNooogaJpZM4LSb5I>
> .
>

twiecki committed:
TST Add thinning.

springcoil commented:
Ahh I never thought of thinning

springcoil commented:
I'll merge this later if it passes
